src/main/scala/controllers/ActionHandler.scala

```diff
@@ -5,35 +5,54 @@ import javax.inject.Inject
 import cats.data.OptionT
 import cats.instances.future._
 import forms.Field
+import forms.validation.{CostItem, CostItemValues}
 import models._
-import play.api.libs.json.JsObject
+import play.api.libs.json.{JsArray, JsDefined, JsObject, Json}
 import play.api.mvc.Result
 import play.api.mvc.Results._
 import services.{ApplicationFormOps, ApplicationOps, OpportunityOps}
 
 import scala.concurrent.{ExecutionContext, Future}
 
-class ActionHandler @Inject()(applications: ApplicationOps, applicationForms: ApplicationFormOps, opportunities: OpportunityOps)(implicit ec: ExecutionContext) {
+class ActionHandler @Inject()(applications: ApplicationOps, applicationForms: ApplicationFormOps, opportunities: OpportunityOps)(implicit ec: ExecutionContext)
+  extends ApplicationResults {
 
   import ApplicationData._
   import FieldCheckHelpers._
 
   def doSave(id: ApplicationId, sectionNumber: Int, fieldValues: JsObject): Future[Result] = {
-    allValuesEmpty(fieldValues) match {
-      case true => applications.deleteSection(id, sectionNumber).map { _ =>
-        Redirect(routes.ApplicationController.show(id))
-      }
-      case false => applications.saveSection(id, sectionNumber, fieldValues).map { _ =>
-        Redirect(routes.ApplicationController.show(id))
-      }
+    sectionTypeFor(sectionNumber) match {
+      case VanillaSection =>
+        allValuesEmpty(fieldValues) match {
+          case true => applications.deleteSection(id, sectionNumber).map { _ =>
+            Redirect(routes.ApplicationController.show(id))
+          }
+          case false => applications.saveSection(id, sectionNumber, fieldValues).map { _ =>
+            Redirect(routes.ApplicationController.show(id))
+          }
+        }
+
+      case CostSection => Future.successful(redirectToOverview(id))
     }
   }
 
   def doComplete(id: ApplicationId, sectionNumber: Int, fieldValues: JsObject): Future[Result] =
-    applications.completeSection(id, sectionNumber, fieldValues).flatMap {
+    sectionTypeFor(sectionNumber) match {
+      case VanillaSection => applications.completeSection(id, sectionNumber, fieldValues).flatMap {
+        case Nil => Future.successful(Redirect(routes.ApplicationController.show(id)))
+        case errs => redisplaySectionForm(id, sectionNumber, JsonHelpers.flatten("", fieldValues), errs)
+      }
 
-      case Nil => Future.successful(Redirect(routes.ApplicationController.show(id)))
-      case errs => redisplaySectionForm(id, sectionNumber, JsonHelpers.flatten("", fieldValues), errs)
+      case CostSection =>
+        applications.getSection(id, sectionNumber).flatMap {
+          case Some(section) =>
+            applications.completeSection(id, sectionNumber, section.answers).flatMap {
+              case Nil => Future.successful(redirectToOverview(id))
+              case errs => redisplaySectionForm(id, sectionNumber, JsonHelpers.flatten("", section.answers), errs)
+            }
+
+          case None => Future.successful(NotFound)
```

Comment:
Why would we throw "Page not found" at the user here? Should we have an error handling page instead?